Can't add `f64 + {float}`

```rust
fn foo(x: f64) {
    let y = x + 1.0;
}
```

```rust
error[E0277]: cannot add `{float}` to `f64`
 --> <source>:2:15
  |
2 |     let y = x + 1.0;
  |               ^ no implementation for `f64 + {float}`
  |
  = help: the trait `Add<{float}>` is not implemented for `f64`
  = help: the following other types implement trait `Add<Rhs>`:
            <&'a f64 as Add<f64>>
            <&f64 as Add<&f64>>
            <f64 as Add<&f64>>
            <f64 as Add>

error: the type `fn(f64, _) -> <f64 as Add<_>>::Output` is not well-formed
 --> <source>:2:15
  |
2 |     let y = x + 1.0;
  |               ^
```

That WF error is strange.

---

Probably also fixes: tests/ui/associated-types/issue-18655.rs

--- a/compiler/rustc_trait_selection/src/solve/canonicalize.rs
+++ b/compiler/rustc_trait_selection/src/solve/canonicalize.rs
@@ -291,7 +291,7 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for Canonicalizer<'_, 'tcx> {
                 if nt != t {
                     return self.fold_ty(nt);
                 } else {
-                    CanonicalVarKind::Ty(CanonicalTyVarKind::Int)
+                    CanonicalVarKind::Ty(CanonicalTyVarKind::Float)
                 }
             }
             ty::Infer(ty::FreshTy(_) | ty::FreshIntTy(_) | ty::FreshFloatTy(_)) => {

--- a/tests/ui/traits/new-solver/float-canonical.rs
+++ b/tests/ui/traits/new-solver/float-canonical.rs
@@ -1,0 +1,8 @@
+// compile-flags: -Ztrait-solver=next
+// check-pass
+
+fn foo(x: f64) {
+    let y = x + 1.0;
+}
+
+fn main() {}

